### PR TITLE
save-demo-agent server

### PR DESCRIPTION
**Pull request by sanyavertolet**
### What's done:
 * Added `/setup` endpoint to pass all the required information for demo run
 * Implemented `SimpleRunner.kt` in order to run the tools

### TODO:
 * Implement tool download from `save-demo`
 * Improve logging

### CODE REVIEW

File: save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/utils/FileUtils.kt [hunk 1]
```diff
@@ -113,6 +114,24 @@ fun Path.requireIsAbsolute(): Path = apply {
     }
 }
 
+actual fun okio.Path.markAsExecutable() {
+    val file = this.toFile().toPath()
+    Files.setPosixFilePermissions(file, Files.getPosixFilePermissions(file) + setOf(
```
Comment (0x6675636b796f75676974687562) [inline save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/utils/FileUtils.kt:119]:
```suggestion
    Files.setPosixFilePermissions(file, Files.getPosixFilePermissions(file) + EnumSet.of(
```

File: save-cloud-common/src/nativeMain/kotlin/com/saveourtool/save/utils/FileUtils.kt [hunk 2]
```diff
@@ -7,11 +7,46 @@ package com.saveourtool.save.utils
...
+import platform.posix.*
 
+import kotlinx.cinterop.UnsafeNumber
 import kotlinx.serialization.serializer
 
 actual val fs: FileSystem = FileSystem.SYSTEM
 
+@OptIn(UnsafeNumber::class)
+actual fun Path.markAsExecutable() {
+    @Suppress("CAST_NEVER_SUCCEEDS") val mode: mode_t = (S_IRUSR or S_IWUSR or S_IXUSR or S_IRGRP or S_IROTH) as mode_t
```
Comment (0x6675636b796f75676974687562) [inline save-cloud-common/src/nativeMain/kotlin/com/saveourtool/save/utils/FileUtils.kt:20]: Are you sure?

This is `rwxr--r--`.

You probably should just read the existing attributes and just merge them with `ugo+x` (`a+x`).

If you can't read the existing attributes using Okio, just grab the default attribute set (0644) and apply the `x` bit (0644 -> 0755).

**Review by 0x6675636b796f75676974687562 — APPROVED** [review]

File: save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/DemoAgentConfig.kt [hunk 3]
```diff
@@ -61,5 +59,5 @@ data class RunConfiguration(
  */
 @Serializable
 data class ServerConfiguration(
-    val port: Int = 23456,
+    val port: Long = 23456L,
```
Comment (0x6675636b796f75676974687562) [inline save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/DemoAgentConfig.kt:62]: This is a TCP port number, right?

But why? `Int` is more than enough.

File: save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/utils/FileUtils.kt (hunk 1 above)
Comment (sanyavertolet) [reply save-cloud-common/src/jvmMain/kotlin/com/saveourtool/save/utils/FileUtils.kt]: Sure, done

File: save-cloud-common/src/nativeMain/kotlin/com/saveourtool/save/utils/FileUtils.kt (hunk 2 above)
Comment (sanyavertolet) [reply save-cloud-common/src/nativeMain/kotlin/com/saveourtool/save/utils/FileUtils.kt:20]: Just moved existing code to make it Multiplatform. Decided to leave as is in order to not to break anything

File: save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/utils/HttpUtils.kt [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                        fs.appendingSink(file, mustExist = hasAlreadyAnyData).buffer().use { it.write(bytes) }
+                        totalBytes.addAndGet(bytes.size.toLong())
+                        logDebug("Received ${bytes.size} ($totalBytes) bytes out of ${httpResponse.contentLength()} bytes from ${httpResponse.request.url}")
+                    }
+                }
+                if (totalBytes.value == 0L) {
+                    error("Downloaded a file from $url but content is empty")
+                }
+            } else {
+                logWarn("Skipping downloading as request is not a success")
```
Comment (0x6675636b796f75676974687562) [inline save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/utils/HttpUtils.kt:55]: Please log the status code and status line (e. g.: `HTTP 418 I'm a teapot`.

File: save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/DemoAgentConfig.kt (hunk 3 above)
Comment (sanyavertolet) [reply save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/DemoAgentConfig.kt:62]: The point is that ktor receives port as `Long`. In order to not to cast it explicitly I've decided to do so.

File: save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/utils/SetupUtils.kt [hunk 5]
```diff
@@ -4,21 +4,80 @@
...
-    downloadDemoFiles(demoUrl, demoConfiguration)
-    executeSetupSh()
+@OptIn(ExperimentalStdlibApi::class)
+suspend fun setupEnvironment(demoUrl: String, demoConfiguration: DemoConfiguration) {
+    logInfo("Setting up the environment...")
+
+    try {
+        downloadDemoFiles(demoUrl, demoConfiguration)
+    } catch (e: IllegalStateException) {
+        logError("Error while downloading files to agent.")
```
Comment (0x6675636b796f75676974687562) [inline save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/utils/SetupUtils.kt:37]: Does it make sense to also `log e.message ?: e.toString()`?

File: save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/utils/SetupUtils.kt [hunk 6]
```diff
@@ -4,21 +4,80 @@
...
 
-// todo: implement setup.sh run
-private fun executeSetupSh() = Unit
+private fun executeSetupSh(setupShName: String = "setup.sh"): ExecutionResult? = setupShName.takeIf {
+    fs.exists(it.toPath())
+}
+    ?.let { setupSh ->
+        setupSh.toPath().markAsExecutable()
+        ProcessBuilder(true, fs).exec(
+            "./$setupSh",
```
Comment (0x6675636b796f75676974687562) [inline save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/utils/SetupUtils.kt:61]:
```suggestion
            "$CWD/$setupSh",
```

**Review by 0x6675636b796f75676974687562 — APPROVED** [review]

File: save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/DemoAgentConfig.kt (hunk 3 above)
Comment (0x6675636b796f75676974687562) [reply save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/DemoAgentConfig.kt:62]: Ah, ok.

File: save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/utils/SetupUtils.kt (hunk 6 above)
Comment (sanyavertolet) [reply save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/utils/SetupUtils.kt:61]: As far as I pass `CWD` to `ProcessBuilder`, it seem to change current working directory to what I pass there (here, it is `CWD`.) Is there any need to do what you want?

File: save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/utils/SetupUtils.kt (hunk 5 above)
Comment (sanyavertolet) [reply save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/utils/SetupUtils.kt:37]: Ok, sure

File: save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/utils/HttpUtils.kt (hunk 4 above)
Comment (sanyavertolet) [reply save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/utils/HttpUtils.kt:55]: Will do it

File: save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/utils/SetupUtils.kt (hunk 5 above)
Comment (sanyavertolet) [reply save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/utils/SetupUtils.kt]: Done

File: save-cloud-common/src/nativeMain/kotlin/com/saveourtool/save/utils/FileUtils.kt (hunk 2 above)
Comment (0x6675636b796f75676974687562) [reply save-cloud-common/src/nativeMain/kotlin/com/saveourtool/save/utils/FileUtils.kt:20]: Ok then. Let's leave as-is.

File: save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/utils/SetupUtils.kt (hunk 6 above)
Comment (sanyavertolet) [reply save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/utils/SetupUtils.kt:61]: Yep, but is it really required? I mean, if `CMD` is `/home`, then the command will look like: `cd /home && /home/setup.sh`

Comment (0x6675636b796f75676974687562) [reply save-demo-agent/src/nativeMain/kotlin/com/saveourtool/save/demo/agent/utils/SetupUtils.kt:61]: Indeed.

Please disregard.

**Pull request merged**